### ISSUE TITLE
"Couldn't load TypeScript from path undefined" error when open .astro file

### ISSUE THREAD

**Issue by privetkuku** [sublimelsp/LSP-astro#12]
Hey folks!
I have an issue with TypeScript in `.astro` files. I am using macOS 12.6 and Sublime Text 4 4126.
So, the problem is that TypeScript does not work in `.astro` files. I have both Astro and LSP-astro packages installed. When I try to open `.astro` file I get this error:
![grafik](https://user-images.githubusercontent.com/114505663/192571410-b9ebc267-5b64-4970-b8b8-288a411e5804.png)
TypeScript work fine in `.tsx` files. I suppose, error is might be because of `pnpm` specifics, but I am not sure. Could you please look into it?
Thanks in advance!

**Comment by deathaxe** [comment]
It appears astro tries to utilize an existing typescript language server. I tried to add `{ST}\DataPackage Storage\LSP-typescript\16.14.0\typescript-language-server\node_modules\typescript-language-server\lib\cli.js` to PATH but it looks the script expects something else.

I am however not familiar enough with TypeScript, VS Code nor the architecture of Astro Language Server to find out what would be required to get that working.

It also appears, Astro ships a plugin for typescript language server, which I have no idea how to setup/install in ST's typescript language server plugin.

Sovling that issue, would require someone of more knowledge about Astro language server.

**Comment by rchl** [comment]
Is `typescript` installed in the project? Servers can sometimes find it automatically if it's installed.

I could look deeper into it but it would help to have a simple (or less simply) project that reproduces the issue.

**Comment by deathaxe** [comment]
I've just opened the [preview.astro](https://github.com/SublimeText/Astro/blob/main/preview.astro) and see the error message as well. Both LSP-astro and LSP-typescript installed. Maybe should find a way to automatically point LSP-astro to LSP-typescript in some way.

**Comment by rchl** [comment]
But aren't astro projects typically node/npm-based? I would kinda expect problems when opening single files like that.

**Comment by deathaxe** [comment]
I have absolutely no idea. Never used Astro beyond getting Astro.sublime-syntax working.

**Comment by rchl** [comment]
Yeah, most likely those need package.json, node_modules and all that shebang.

Additionally the original comment is about using non-standard pnpm which uses some virtual/redirected file system or something so node_modules might not be visible to the server like with npm.

Thus I'm not touching it without a repo that reproduces the issue.

**Comment by privetkuku** [comment]
Hey folks, thanks for looking into it!
I put up small repo, where you may observe the issue - https://github.com/privetkuku/lsp-astro-sublime-issue.
TypeScript is used in the project and it's working for typescript files (just try adding dummy file like `test.ts` in the project and trigger some error).
Hope that helps!

**Comment by rchl** [comment]
Haven't looked at it yet but just a note about *.ts files... This server is not set up to open *.ts files so it working there is not that relevant as it's a different server that handles those (LSP-typescript).

**Comment by rchl** [comment]
For now you can add `typescript` dependency in your project and then set full path to it in the LSP-astro settings.

Add this (replace `[...]` with full path):
```js
	"initializationOptions": {
		"typescript.serverPath": "[...]/lsp-astro-sublime-issue/node_modules/typescript/lib/tsserverlibrary.js",
	},
```

I guess this package would have to implement all this custom logic to lookup typescript in the project or fall back to bundled typescript like the VSCode extension does.

**Comment by privetkuku** [comment]
Thanks Rafał, it works!
![grafik](https://user-images.githubusercontent.com/114505663/193031559-40af35e8-2735-4dca-b566-e42fd736ab19.png)
Side question, do I understand corretly that `tsserver` is built from`tsserverlibrary.js`?

**Comment by deathaxe** [comment]
Wouldn't it be enough to add typescript package to the list of dependencies and point initializationOptions to it?

```json
	"initializationOptions": {
		"typescript.serverPath": "$server_directory_path/node_modules/typescript/lib/tsserverlibrary.js",
	},
```

**Comment by rchl** [comment]
Ideally LSP-astro should lookup and use `typescript` from the project first and only fallback to bundled otherwise.

This is because different versions of typescript can produce different results so it would be confusing to see different results from LSP and different when building the project.

Though it would be acceptable to use bundled typescript as a quickfix.

**Comment by deathaxe** [comment]
Same would apply to LSP-typescript, wouldn't it? A project could target any version of Typescript, while LSP-typescript always uses TS 4.8.3 at the moment.

**Comment by rchl** [comment]
No, LSP-typescript uses typescript from workspace if available. The difference there is that the typescript-language-server is the one that implements the logic to lookup typescript.

LSP-volar is an example of a package that does typescript lookup itself.

https://github.com/sublimelsp/LSP-volar/blob/eb60ac7cf2cdd358690017e51830973ac74e5365/plugin.py#L47-L50

**Comment by deathaxe** [comment]
Looks pretty simple. I wouldn't run a node script just to resolve a path though.

**Comment by privetkuku** [comment]
> Wouldn't it be enough to add typescript package to the list of dependencies and point initializationOptions to it?
> 
> ```json
> 	"initializationOptions": {
> 		"typescript.serverPath": "$server_directory_path/node_modules/typescript/lib/tsserverlibrary.js",
> 	},
> ```

Hey, where is `$server_directory_path` variable get assigned?

**Comment by rchl** [comment]
In the `lsp_utils` dependency: https://github.com/sublimelsp/lsp_utils/blob/467636293a0e92e5797a6ad98423b70b97b42f80/st3/lsp_utils/npm_client_handler.py#L69-L73

**Comment by deathaxe** [comment]
The issue is resolved by LSP-astro v1.2.0+

**Comment by joshmedeski** [comment]
I ran into the same problem while using the Astro LSP installed with Mason on Neovim. I installed TypeScript onto the project and it started working, thanks!
